Learn to parse `a as usize < b`

estebank opened the pull request:
Parsing `a as usize > b` always works, but `a as usize < b` was a
parsing error because the parser would think the `<` started a generic
type argument for `usize`. The parser now attempts to parse as before,
and if a DiagnosticError is returned, try to parse again as a type with
no generic arguments. If this fails, return the original
`DiagnosticError`.

Fix #22644.

estebank committed:
Learn to parse `a as usize < b`

Parsing `a as usize > b` always works, but `a as usize < b` was a
parsing error because the parser would think the `<` started a generic
type argument for `usize`. The parser now attempts to parse as before,
and if a DiagnosticError is returned, try to parse again as a type with
no generic arguments. If this fails, return the original
`DiagnosticError`.

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

oli-obk commented on src/libsyntax/parse/parser.rs:
```diff
@@ -2798,8 +2823,38 @@ impl<'a> Parser<'a> {
+                        };
+                        let path = TyKind::Path(None, path);
+                        let span = lo.to(self.prev_span);
+                        let rhs = P(Ty { node: path, span: span, id: ast::DUMMY_NODE_ID });
```
Is it possible to use `mk _ty` here? Or is the  `DUMMY_NODE_ID` intentional?

estebank replied:
`mk_ty` is part of `TyCtxt` which isn't available yet. This was extracted from `parse_ty_common` in order to see if this would work at all. I could modify `parse_ty_common` for this case, but I felt it would muddy the logic of that method.

petrochenkov commented:
I'd be okay with this if the unbounded lookahead and parser state saving/restoring machinery were used only for reporting better diagnostics and fix-it suggestions, not for actually accepting the code.

Mark-Simulacrum commented:
Looks like there are some test failures in Travis: https://travis-ci.org/rust-lang/rust/jobs/241424157#L1853, though addressing @petrochenkov's concern first seems like a good idea.

estebank commented:
> I'd be okay with this if the unbounded lookahead and parser state saving/restoring machinery were used only for reporting better diagnostics, recovery and fix-it suggestions, not for actually accepting the code.

Should `a as usize < b` be accepted? I understand the concern, and would heavily lean towards your opinion on all other situations, but in this case in particular it is adding a very specific edge case to the grammar which IMO is a big ergonomic win.

Having said that, what would be your preferred output if this were used only for diagnostics? Something like this?

```
error: expected one of `!`, `(`, `+`, `,`, `::`, `<`, or `>`, found `;`
 --> <anon>:6:19
  |
6 |     a as usize < b;
  |                   ^ expected one of 7 possible tokens here
  help: may be you want to compare the casted value?
  |     (a as usize) < b;
```

petrochenkov commented:
@estebank 
>Having said that, what would be your preferred output if this were used only for diagnostics?

A specially crafted error would be the best, IMO.
```
error: `<` is interpreted as a start of generic arguments for `usize`, not comparison
 --> <anon>:6:19
  |
6 |     a as usize < b;
  |                ^ <some label, not sure what to write here>
  help: if you want to compare the casted value then write
  |     (a as usize) < b;
```
After this is reported as a non-fatal error, AST is build as if `a as usize < b` were legal and meant `(a as usize) < b`, e.g. most likely error recovery is performed.

petrochenkov commented:
@estebank 
>it is adding a very specific edge case to the grammar which IMO is a big ergonomic win.

It's an edge case, but it has large effect on properties of the grammar as a whole.
If syntactic disambiguation between types and expressions is going to be done, it should probably be done in more principled fashion, discussed properly and maybe modeled in something like https://github.com/nikomatsakis/rustypop
BTW, the [Const generics RFC](https://github.com/rust-lang/rfcs/pull/2000) is related, first it break the assumption this PR does that if `b` is not a type, then it's the rhs argument for comparison, second it creates more demand for general disambiguation between types and expressions.

petrochenkov commented:
>`(a as usize) < b`

`a as (usize) < b` works too and is more "local", but arguably more surprising, maybe it can be recommended instead? (I'm not sure)

estebank committed:
Suggest non-ambiguous comparison after cast

```
warning: `<` is interpreted as a start of generic arguments for `usize`, not comparison
  --> $DIR/issue-22644.rs:16:33
   |
16 |     println!("{}", a as usize < b);
   |                                 ^ expected one of `!`, `(`, `+`, `,`, `::`, or `>` here
   |
help: if you want to compare the casted value then write
   |     println!("{}", (a as usize) < b);
```

estebank committed:
Fix affected tests

estebank commented:
> It's an edge case, but it has large effect on properties of the grammar as a whole.  
> If syntactic disambiguation between types and expressions is going to be done, it should probably be done in more principled fashion, discussed properly and maybe modeled in something like https://github.com/nikomatsakis/rustypop

I understand and completely agree.

> a as (usize) < b works too and is more "local", but arguably more surprising, maybe it can be recommended instead? (I'm not sure)

I think I prefer the former, as it'd be less surprising for a newcomer.

I got something working which accepts the code as proposed but also emits a warning suggesting a correction.

I should probably change the code that emits the warning before we merge, as right now [I am modifying the existing error in place](https://github.com/rust-lang/rust/pull/42578/files#diff-da9d34ca1d0f6beee2838cf02e07345cR2853), which is probably not what we want.

This is how it looks like right now:

```
warning: `<` is interpreted as a start of generic arguments for `usize`, not comparison
  --> $DIR/issue-22644.rs:16:33
   |
16 |     println!("{}", a as usize < b);
   |                                 ^ expected one of `!`, `(`, `+`, `,`, `::`, or `>` here
   |
help: if you want to compare the casted value then write:
   |     println!("{}", (a as usize) < b);
```

It probably should be closer to:

```
warning: `<` is interpreted as a start of generic arguments for `usize`, not comparison
  --> $DIR/issue-22644.rs:16:33
   |
16 |     println!("{}", a as usize < b);
   |                               - ^ interpreted as generic argument
   |                               |
   |                               not interpreted as comparison
help: if you want to compare the casted value then write:
   |     println!("{}", (a as usize) < b);
```

bjorn3 commented on src/libsyntax/parse/parser.rs:
```diff
@@ -438,6 +441,8 @@ fn dummy_arg(span: Span) -> Arg {
     Arg { ty: P(ty), pat: pat, id: ast::DUMMY_NODE_ID }
 }
 
+type RewindPoint = (token::Token, Span, Option<Span>, Span, TokenCursor, Vec<TokenType>);
```
Could you make this a struct with field names?

estebank committed:
Change `<` interpreted as generic arg start warning

```
warning: `<` is interpreted as a start of generic arguments for `usize`, not a comparison
  --> $DIR/issue-22644.rs:16:33
   |
16 |     println!("{}", a as usize < b);
   |                               - ^ interpreted as generic argument
   |                               |
   |                               not interpreted as comparison
   |
help: if you want to compare the casted value then write:
   |     println!("{}", (a as usize) < b);
```

nikomatsakis commented:
As long as we're not changing the grammar, I'm happy if @petrochenkov is happy =)

nikomatsakis commented:
r? @petrochenkov

estebank commented:
@nikomatsakis just to be clear, the code as is _accepts_ it under the assumption that it is correct and only emits a `warn` diagnostic. If this is the _only_ error, it will generate a valid binary. Is it ok to accept this code relying on people to be annoyed enough by warnings to fix their code?

bjorn3 commented on src/libsyntax/parse/parser.rs:
```diff
@@ -1819,7 +1850,8 @@ impl<'a> Parser<'a> {
 
             // Parse types, optionally.
-            let parameters = if self.eat_lt() {
+            let parameters = if self.is_lt() && !dont_parse_generics {
```
`if !dont_parse_generics && self.eat_lt() {`

`&&` doesnt run second arg when first arg is false.

petrochenkov replied:
👍 
This also makes `fn is_lt` unnecessary.

nikomatsakis commented:
@estebank 

> the code as is accepts it under the assumption that it is correct and only emits a warn diagnostic

Accepts what exactly? And what do you mean by "as is"? You mean, in the PR, or in master?

estebank commented:
@nikomatsakis The PR accepts `a as b < c` as valid code if it can parse `b` as a type without type arguments, showing a warning, not an error to change the code.

petrochenkov commented on src/libsyntax/parse/parser.rs:
```diff
@@ -1800,7 +1829,9 @@ impl<'a> Parser<'a> {
     /// - `a::b<T,U>::c(V) -> W`
     /// - `a::b<T,U>::c(V)`
-    pub fn parse_path_segments_without_colons(&mut self) -> PResult<'a, Vec<PathSegment>> {
+    pub fn parse_path_segments_without_colons(&mut self, dont_parse_generics: bool)
```
Nit: `dont_parse_generics: bool` -> `parse_generics: bool`

estebank replied:
Done.

petrochenkov commented on src/libsyntax/parse/parser.rs:
```diff
@@ -2798,8 +2830,62 @@ impl<'a> Parser<'a> {
+                                           ExprKind::Cast(lhs, rhs), ThinVec::new());
+                    }
+                    Err(mut err) => {
+                        // Rewind to before attempting to parse the type with generics, to get
```
Could you move the error handling logic into a separate function?

estebank replied:
Done.

petrochenkov commented on src/libsyntax/parse/parser.rs:
```diff
@@ -2798,8 +2830,62 @@ impl<'a> Parser<'a> {
+                                let msg = format!("`<` is interpreted as a start of generic \
+                                                   arguments for {}, not a comparison",
+                                                  warn_message);
+                                let mut warn = self.sess.span_diagnostic.struct_span_warn(sp, &msg);
```
`struct_span_err` please.

estebank replied:
Done.

petrochenkov commented on src/libsyntax/parse/parser.rs:
```diff
@@ -2798,8 +2830,62 @@ impl<'a> Parser<'a> {
+                                warn.emit();
+                                path
+                            }
+                            Err(mut path_err) => {
```
Could you add a test for this case as well?

nikomatsakis commented:
> The PR accepts `a as b < c` as valid code if it can parse b as a type without type arguments, showing a warning, not an error to change the code.

Huh, I'm confused. After all, you can do `let x = foo as Box<Bar>`. How do we know that `b < c` should *not* be parsed as a type?

estebank committed:
Review comments

- generate error instead of warning
- remove `RewindPoint` and just keep a copy of `Parser` to rewind state.
- `dont_parse_generics: bool` -> `parse_generics: bool`
- remove `eat_lt`
- move error handling code to separate method

estebank commented on src/libsyntax/parse/parser.rs:
```diff
@@ -438,6 +441,15 @@ fn dummy_arg(span: Span) -> Arg {
     Arg { ty: P(ty), pat: pat, id: ast::DUMMY_NODE_ID }
 }
 
+struct RewindPoint {
```
Removed. Using `Parser.clone()` and `mem::replace()` instead of `RewindPoint`.

petrochenkov commented:
r=me with [stray note](https://github.com/rust-lang/rust/pull/42578#commitcomment-22556289) removed.

estebank commented:
@petrochenkov done

@bors r=petrochenkov

bors commented:
:pushpin: Commit ad260ff has been approved by `petrochenkov`

<!-- @bors r=petrochenkov ad260ffc884477686e8f6f52c97a417bf99e2f19 -->

nikomatsakis commented:
@bors r-

nikomatsakis commented:
I'd like to know if this is changing the grammar that is accepted. I feel like I still don't understand quite what's happening here. =)

nikomatsakis commented:
In particular, do we now compile more programs than we did before? If so, what set of programs?

estebank commented:
@nikomatsakis The first commit of this PR changed the accepted grammar and only emitted a warning. I changed it back to *not* have any changes in the grammar and emit an error. The parser state changes slightly after finding this case in order to continue parsing successfully the rest of the file (before this PR the rest of the file would not get any diagnostics). No new set of programs will be accepted, the new behaviors are:

 - a more specific error for the case `a as b < c`
 - continued parsing of the rest of the file after encountering `a as b < c`

nikomatsakis commented:
@estebank perfect =)

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit ad260ff has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis ad260ffc884477686e8f6f52c97a417bf99e2f19 -->

bors commented:
:hourglass: Testing commit ad260ff with merge 44eeb21...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.3598), [status-travis](https://travis-ci.org/rust-lang/rust/builds/243520596?utm_source=github_status&utm_medium=notification)
Approved by: nikomatsakis
Pushing 44eeb2109b249e7496ca0897ce509ced2e3b6c81 to master...

hertz4 commented:
One last thing: `<<` isn't covered here.
```rust
error: `<` is interpreted as a start of generic arguments for `u16`, not a comparison                                          
 --> /tmp/angle.rs:4:29                                                                                                        
  |                                                                                                                            
4 |     println!("{}", x as u16 << 4);                                                                                         
  |                    -------- ^^ - interpreted as generic arguments                                                          
  |                    |        |                                                                                              
  |                    |        not interpreted as comparison                                                                  
  |                    help: try comparing the casted value: `(x as u16)`       
```

petrochenkov commented:
@hertz4 
`<<` is fixed on beta/nightly

```
error: `<` is interpreted as a start of generic arguments for `u16`, not a shift
 --> src/main.rs:2:29
  |
2 |     println!("{}", x as u16 << 4);
  |                    -------- ^^ - interpreted as generic arguments
  |                    |        |
  |                    |        not interpreted as shift
  |                    help: try shifting the casted value: `(x as u16)`
```
